src/charm.py

```diff
@@ -1,71 +1,77 @@
+
 
 class Operator(CharmBase):
     def __init__(self, *args):
         super().__init__(*args)
 
-        self.log = logging.getLogger()
+        self.log = logging.getLogger(__name__)
 
-        if not self.model.unit.is_leader():
-            self.log.info("Not a leader, skipping set_pod_spec")
-            self.model.unit.status = ActiveStatus()
+        if not self.unit.is_leader():
+            self.log.info("Not a leader, skipping configuration")
+            self.unit.status = ActiveStatus()
             return
 
-        self.image = OCIImageResource(self, "oci-image")
-
         try:
             self.interfaces = get_interfaces(self)
         except NoVersionsListed as err:
-            self.model.unit.status = WaitingStatus(str(err))
+            self.unit.status = WaitingStatus(str(err))
             return
         except NoCompatibleVersions as err:
-            self.model.unit.status = BlockedStatus(str(err))
+            self.unit.status = BlockedStatus(str(err))
             return
         else:
-            self.model.unit.status = ActiveStatus()
+            self.unit.status = ActiveStatus("")
 
-        self.framework.observe(self.on.install, self.set_pod_spec)
-        self.framework.observe(self.on.upgrade_charm, self.set_pod_spec)
-        self.framework.observe(self.on.config_changed, self.set_pod_spec)
-        self.framework.observe(self.on["mysql"].relation_changed, self.set_pod_spec)
+        self.framework.observe(self.on.install, self.update_container)
+        self.framework.observe(self.on.upgrade_charm, self.update_container)
+        self.framework.observe(self.on.config_changed, self.update_container)
+        self.framework.observe(self.on["mysql"].relation_changed, self.update_container)
 
         self.framework.observe(self.on["grpc"].relation_changed, self.send_info)
 
-    def send_info(self, event):
+        # run service patcher after the charm hooks have run
+        # so that config['port'] is set correctly
+        self.service_patcher = KubernetesServicePatch(
+            self, [("grpc-api", int(self.config["port"]))]
+        )
+
+    def send_info(self, _):
         if self.interfaces["grpc"]:
             self.interfaces["grpc"].send_data(
                 {
                     "service": self.model.app.name,
-                    "port": self.model.config["port"],
+                    "port": self.config["port"],
                 }
             )
 
-    def set_pod_spec(self, event):
-        try:
-            image_details = self.image.fetch()
-        except OCIImageResourceError as e:
-            self.model.unit.status = e.status
-            self.log.info(e)
-            return
+    def update_container(self, _):
+        # update port in service patcher
+        self.service_patcher.service = self.service_patcher._service_object(
+            [("grpc-api", int(self.config["port"]))]
+        )
+        self.service_patcher._patch(None)
```

Comment:
Slightly ugly hack to work around the fact the KubernetesServicePatch does not allow dynamic ports.

In general, I doubt we really need to change the port with charm config, but I've preserved the original charm behaviour.

Reply:
^^ seconding this, i think this is how other operators also do it nowadays.